### PR TITLE
Add SELF_DESTRUCT env variable to process self-destructions in the background

### PR DESCRIPTION
**Pull request by ClearlyClaire**
This is an alternative to #26438: because we already require Sidekiq to be running, why not perform the deletion in Sidekiq itself?

This adds the `SELF_DESTRUCT` environment variable, which, when it is set to the proper value, switches Mastodon to self-destruct mode, stopping processing API and ActivityPub requests, and deleting accounts in a background process, hopefully fast enough to not take ages, and slow enough to not overwhelm Sidekiq's Redis.

To get the proper value for the `SELF_DESTRUCT` environment variable, you need to call `tootctl self-destruct`, which can also give you information about the progress of the self-destruct process:
```console
[mastodon@mastodev live]$ ./bin/tootctl self-destruct
Type in the domain of the server to confirm: mastodev.sitedethib.com
This operation WILL NOT be reversible. It can also take a long time.
While the data won't be erased locally, the server will be in a BROKEN STATE afterwards.
A running Sidekiq process is required. Do not shut it down until queues clear.
Are you sure you want to proceed? yes
To switch Mastodon to self-destruct mode, add the following line to your `.env.production` and restart all Mastodon processes:
  SELF_DESTRUCT=BAhJIhxtYXN0b2Rldi5zaXRlZGV0aGliLmNvbQY6BkVU--6a49f917d92fc9e3e0bcebf68b61cf431bc97eba

You can re-run this command to see the state of the self-destruct process.
[mastodon@mastodev live]$ echo 'SELF_DESTRUCT=BAhJIhxtYXN0b2Rldi5zaXRlZGV0aGliLmNvbQY6BkVU--6a49f917d92fc9e3e0bcebf68b61cf431bc97eba' >> .env.production
[mastodon@mastodev live]$ sudo systemctl restart mastodon-{web,sidekiq,streaming}
[mastodon@mastodev live]$ ./bin/tootctl self-destruct
Self-destruct mode is already enabled for this Mastodon server
7608 accounts are still pending deletion.
[mastodon@mastodev live]$ ./bin/tootctl self-destruct
Self-destruct mode is already enabled for this Mastodon server
7558 accounts are still pending deletion.
[mastodon@mastodev live]$ ./bin/tootctl self-destruct
Self-destruct mode is already enabled for this Mastodon server
Deletion notices are still being processed
[mastodon@mastodev live]$ ./bin/tootctl self-destruct
Self-destruct mode is already enabled for this Mastodon server
At least one delivery attempt for each deletion notice has been made, but some have failed and are scheduled for retry
```

It also allows users to fetch their data, but nothing else. If this is not desired, `mastodon-web` should be stopped rather than restarted with `SELF_DESTRUCT`.

- [x] allows logging in and logging out
- [x] allows viewing sessions, login history, change password and 2FA
- [x] allows exporting CSV lists and archive export when logged in
- [x] allows access to the Sidekiq web UI and PGHero when logged in as an admin
- [x] returns 410 for every other path
- [x] only schedule the `SelfDestructScheduler` in the self-destruct mode
- [ ] schedule only the `SelfDestructScheduler` and a couple other cleanup schedulers when in self-destruct mode
- [x] show admins the progress of the self-destruct process (done through `tootctl self-destruct` output)

image
image

### CODE REVIEW

File: lib/mastodon/cli/main.rb [hunk 1]
```diff
@@ -92,55 +91,37 @@ def self_destruct
...
+      prompt.warn('A running Sidekiq process is required. Do not shut it down until queues clear.')
 
-      Account.local.without_suspended.find_each { |account| delete_account.call(account) }
-      Account.local.suspended.joins(:deletion_request).find_each { |account| delete_account.call(account) }
+      exit(1) if prompt.no?('Are you sure you want to proceed?')
 
-      prompt.ok("Queued #{inboxes.size * processed} items into Sidekiq for #{processed} accounts#{dry_run_mode_suffix}")
-      prompt.ok('Wait until Sidekiq processes all items, then you can shut everything down and delete the data')
+      self_destruct_value = Rails.application.message_verifier('self-destruct').generate(Rails.configuration.x.local_domain)
+      prompt.ok('To switch Mastodon to self-destruct mode, add the following line to your `.env.production` and restart all Mastodon processes:')
```
Comment (renchap) [inline lib/mastodon/cli/main.rb:122]: What about people not using `.env.production`? I assume they will know they need to change the env var using the method they use for the other env vars, but maybe we could make it more explicit?

**Review by renchap — CHANGES_REQUESTED** [review]

File: app/controllers/settings/login_activities_controller.rb [hunk 2]
```diff
@@ -1,6 +1,9 @@
 # frozen_string_literal: true
 
 class Settings::LoginActivitiesController < Settings::BaseController
+  skip_before_action :check_self_destruct!
+  skip_before_action :require_functional!
```
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/controllers/settings/login_activities_controller.rb:5]: Yeah, suspended accounts are considered non-functional, and all accounts will be marked as suspended sooner or later, but can still benefit from this feature.

File: lib/mastodon/cli/main.rb [hunk 3]
```diff
@@ -92,55 +91,37 @@ def self_destruct
...
-
-          account.suspend!(block_email: false)
-        end
+      exit(1) unless prompt.ask('Type in the domain of the server to confirm:', required: true) == Rails.configuration.x.local_domain
 
-        processed += 1
-      end
+      prompt.warn('This operation WILL NOT be reversible. It can also take a long time.')
+      prompt.warn('While the data won\'t be erased locally, the server will be in a BROKEN STATE afterwards.')
+      prompt.warn('A running Sidekiq process is required. Do not shut it down until queues clear.')
```
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply lib/mastodon/cli/main.rb:117]: I am not sure this is worthwhile, as this is information that may be difficult to keep up to date. In contrast, re-running the command will give the admin the state of the self-destruct process. This is stated at the end of the command, but maybe we could state it earlier.

File: config/sidekiq.yml [hunk 4]
```diff
@@ -62,3 +63,7 @@
       interval: 30 minutes
       class: Scheduler::SoftwareUpdateCheckScheduler
       queue: scheduler
+    self_destruct_scheduler:
```
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply config/sidekiq.yml]: Changed it to do it dynamically, and also remove some costly schedulers while we're at it!

File: config/initializers/sidekiq.rb [hunk 5]
```diff
@@ -37,3 +37,29 @@
   config.reaper_timeout  = 150
   config.lock_ttl        = 50.days.to_i
 end
+
+begin
+  # Start the self-destruct scheduler if needed
+  value = ENV.fetch('SELF_DESTRUCT', nil)
+  if value.present? && Rails.application.message_verifier('self-destruct').verify(value) == ENV['LOCAL_DOMAIN']
```
Comment (renchap) [inline config/initializers/sidekiq.rb:44]: Is there a way to re-use the helper here, or have a shared location used here and in the helper?

File: lib/mastodon/cli/main.rb (hunk 1 above)
Comment (ClearlyClaire) [reply lib/mastodon/cli/main.rb]: I reworded this comment to:
> To switch Mastodon to self-destruct mode, add the following variable to your evironment (e.g. by adding a line to your `.env.production`) and restart all Mastodon processes:

does that work better?

File: lib/mastodon/cli/main.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (renchap) [reply lib/mastodon/cli/main.rb:117]: No issue with this, but maybe reword "a sidekiq process". Because if you run multiple sidekiq processes, each processing different queues, then you need to keep them all active.

**Review by renchap — APPROVED** [review]
